src/Build/BackEnd/Node/MSBuildClient.cs

```diff
@@ -0,0 +1,661 @@
+        /// <param name="dllLocation">Location of dll file to launch the server process if needed.
+        /// Empty if executable is msbuild.exe and not empty if dotnet.exe.</param>
+        public MSBuildClient(string msbuildLocation, string exeLocation, string dllLocation)
+        {
+            ServerEnvironmentVariables = new();
+            _exitResult = new();
+
+            // dll & exe locations
+            _msBuildLocation = msbuildLocation;
+            _exeLocation = exeLocation;
+            _dllLocation = dllLocation;
+
+            // Client <-> Server communication stream
+            _handshake = GetHandshake();
+            _pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuildServer-" + _handshake.ComputeHash());
+            _nodeStream = new NamedPipeClientStream(".", _pipeName, PipeDirection.InOut, PipeOptions.Asynchronous
+#if FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                                                                         | PipeOptions.CurrentUserOnly
+#endif
+            );
+
+            _packetMemoryStream = new MemoryStream();
+            _binaryWriter = new BinaryWriter(_packetMemoryStream);
+
+            // Packet pump block
+            _receivedPacketsQueue = new ConcurrentQueue<INodePacket>();
+            _packetReceivedEvent = new AutoResetEvent(false);
+            _terminatePacketPumpEvent = new ManualResetEvent(false);
+            _packetPumpShutdownEvent = new ManualResetEvent(false);
+            _packetFactory = new NodePacketFactory();
+            _sharedReadBuffer = InterningBinaryReader.CreateSharedBuffer();
+
+            (this as INodePacketFactory).RegisterPacketHandler(NodePacketType.ServerNodeConsoleWrite, ServerNodeConsoleWrite.FactoryForDeserialization, this);
+            (this as INodePacketFactory).RegisterPacketHandler(NodePacketType.ServerNodeBuildResult, ServerNodeBuildResult.FactoryForDeserialization, this);
+        }
+
+        /// <summary>
+        /// Orchestrates the execution of the build on the server,
+        /// responsible for client-server communication.
+        /// </summary>
+        /// <param name="commandLine">The command line to process. The first argument
+        /// on the command line is assumed to be the name/path of the executable, and
+        /// is ignored.</param>
+        /// <returns>A value of type <see cref="MSBuildClientExitResult"/> that indicates whether the build succeeded,
+        /// or the manner in which it failed.</returns>
+        public MSBuildClientExitResult Execute(string commandLine)
+        {
+            string serverRunningMutexName = $@"Global\server-running-{_pipeName}";
+            string serverBusyMutexName = $@"Global\server-busy-{_pipeName}";
+
+            // Start server it if is not running.
+            bool serverWasAlreadyRunning = ServerNamedMutex.WasOpen(serverRunningMutexName);
+            if (!serverWasAlreadyRunning)
+            {
+                if (!LaunchMSBuildServer())
```

Comment:
nit: You can combine this `if` statement with the previous one (`!serverWasAlreadyRunning && !LaunchMSBuildServer()`). C# implements short-circuit evaluation, so the `LaunchMSBuildServer` will not be invoked when the left part is evaluated as `false`.